crates/turborepo-lib/src/package_manager/mod.rs

```diff
@@ -127,43 +128,127 @@ impl Globs {
         })
     }
 
-    pub fn test(&self, root: &Path, target: PathBuf) -> Result<bool> {
-        let search_value = target
-            .strip_prefix(root)?
-            .to_str()
-            .ok_or_else(|| anyhow!("The relative path is not UTF8."))?;
+    pub fn test(
+        &self,
+        root: &AbsoluteSystemPath,
+        target: &AbsoluteSystemPath,
+    ) -> Result<bool, Error> {
+        let search_value = root.anchor(target)?;
 
-        let includes = self.inclusions.is_match(search_value);
-        let excludes = self.exclusions.is_match(search_value);
+        let includes = self.inclusions.is_match(&search_value);
+        let excludes = self.exclusions.is_match(&search_value);
 
         Ok(includes && !excludes)
     }
 }
 
+#[derive(Debug, Error)]
+pub struct MissingWorkspaceError {
+    package_manager: PackageManager,
+}
+
+#[derive(Debug, Error)]
+pub struct NoPackageManager;
+
+impl NoPackageManager {
+    // TODO: determine how to thread through user-friendly error message and apply
+    // our UI
+    pub fn ui_display(&self, ui: &UI) -> String {
+        let url =
+            ui.apply(UNDERLINE.apply_to("https://nodejs.org/api/packages.html#packagemanager"));
+        format!(
+            "We did not find a package manager specified in your root package.json. Please set \
+             the \"packageManager\" property in your root package.json ({url}) or run `npx \
+             @turbo/codemod add-package-manager` in the root of your monorepo."
+        )
+    }
+}
+
+impl Display for NoPackageManager {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "We did not find a package manager specified in your root package.json. \
+        Please set the \"packageManager\" property in your root package.json (https://nodejs.org/api/packages.html#packagemanager) \
+        or run `npx @turbo/codemod add-package-manager` in the root of your monorepo.")
+    }
+}
+
+impl Display for MissingWorkspaceError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let err = match self.package_manager {
+            PackageManager::Pnpm | PackageManager::Pnpm6 => {
+                "pnpm-workspace.yaml: no packages found. Turborepo requires pnpm workspaces and \
+                 thus packages to be defined in the root pnpm-workspace.yaml"
+            }
+            PackageManager::Yarn | PackageManager::Berry => {
+                "package.json: no workspaces found. Turborepo requires Yarn workspaces to be \
+                 defined in the root package.json"
+            }
+            PackageManager::Npm => {
+                "package.json: no workspaces found. Turborepo requires npm workspaces to be \
+                 defined in the root package.json"
+            }
+        };
+        write!(f, "{}", err)
+    }
+}
+
+impl From<&PackageManager> for MissingWorkspaceError {
```

Comment:
Minor detail, but we could definitely derive `Copy` for `PackageManager` since it's essentially the size of a u8

Reply:
I gave this a quick try, and it's less helpful than one might hope. The method where this conversion is used is currently defined on `&self`, meaning that the conversion needs `*self` if you want to take advantage of the copy. Given the size of the type, we could define methods on `self` instead, but I'm not sure that communicates what we want.